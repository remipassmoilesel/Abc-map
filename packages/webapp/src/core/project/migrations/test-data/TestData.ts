/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import * as fs from 'fs';
import { MigrationProject } from '../typings';
import { ProjectHelper, Zipper } from '@abc-map/shared';
import { AbcProjectManifest010 } from '../dependencies/010-project-types';
import { AbcProjectManifest030 } from '../dependencies/030-project-types';
import { AbcProjectManifest020 } from '../dependencies/020-project-types';
import { AbcProjectManifest110 } from '../dependencies/110-project-types';
import { AbcProjectManifest090 } from '../dependencies/090-project-types';
import { AbcProjectManifest040 } from '../dependencies/040-project-types';
import { AbcProjectManifest050 } from '../dependencies/050-project-types';
import { AbcProjectManifest060 } from '../dependencies/060-project-types';
import { AbcProjectManifest070 } from '../dependencies/070-project-types';
import { AbcProjectManifest080 } from '../dependencies/080-project-types';
import { AbcProjectManifest100 } from '../dependencies/100-project-types';
import { AbcProjectManifest130 } from '../dependencies/130-project-types';

export class TestData {
  public static async project01(): Promise<MigrationProject<AbcProjectManifest010>> {
    return this.readFile(`${__dirname}/project-0.1.abm2`) as Promise<MigrationProject<AbcProjectManifest010>>;
  }

  public static async project020(): Promise<MigrationProject<AbcProjectManifest020>> {
    return this.readFile(`${__dirname}/project-0.2.0.abm2`) as Promise<MigrationProject<AbcProjectManifest020>>;
  }

  public static async project030(): Promise<MigrationProject<AbcProjectManifest030>> {
    return this.readFile(`${__dirname}/project-0.3.0.abm2`) as Promise<MigrationProject<AbcProjectManifest030>>;
  }

  public static async project040(): Promise<MigrationProject<AbcProjectManifest040>> {
    return this.readFile(`${__dirname}/project-0.4.0.abm2`) as Promise<MigrationProject<AbcProjectManifest040>>;
  }

  public static async project050(): Promise<MigrationProject<AbcProjectManifest050>> {
    return this.readFile(`${__dirname}/project-0.5.0.abm2`) as Promise<MigrationProject<AbcProjectManifest050>>;
  }

  public static async project060(): Promise<MigrationProject<AbcProjectManifest060>> {
    return this.readFile(`${__dirname}/project-0.6.0.abm2`) as Promise<MigrationProject<AbcProjectManifest060>>;
  }

  public static async project070(): Promise<MigrationProject<AbcProjectManifest070>> {
    return this.readFile(`${__dirname}/project-0.7.0.abm2`) as Promise<MigrationProject<AbcProjectManifest070>>;
  }

  public static async project080(): Promise<MigrationProject<AbcProjectManifest080>> {
    return this.readFile(`${__dirname}/project-0.8.0.abm2`) as Promise<MigrationProject<AbcProjectManifest080>>;
  }

  public static async project090(): Promise<MigrationProject<AbcProjectManifest090>> {
    return this.readFile(`${__dirname}/project-0.9.0.abm2`) as Promise<MigrationProject<AbcProjectManifest090>>;
  }

  public static async project100(): Promise<MigrationProject<AbcProjectManifest100>> {
    return this.readFile(`${__dirname}/project-1.0.0.abm2`) as Promise<MigrationProject<AbcProjectManifest100>>;
  }

  public static async project110(): Promise<MigrationProject<AbcProjectManifest110>> {
    return this.readFile(`${__dirname}/project-1.1.0.abm2`) as Promise<MigrationProject<AbcProjectManifest110>>;
  }

  public static async project110_public(): Promise<MigrationProject<AbcProjectManifest110>> {
    return this.readFile(`${__dirname}/project-1.1.0_public.abm2`) as Promise<MigrationProject<AbcProjectManifest110>>;
  }

  public static async project130(): Promise<MigrationProject<AbcProjectManifest130>> {
    return this.readFile(`${__dirname}/project-1.3.0.abm2`) as Promise<MigrationProject<AbcProjectManifest130>>;
  }

  public static fakeProject<T>(version: string): T {
    return { metadata: { version } } as unknown as T;
  }

  private static async readFile(path: string): Promise<MigrationProject<unknown>> {
    const archive = fs.readFileSync(path);
    const files = await Zipper.forNodeJS().unzip(archive);
    const manifest = await ProjectHelper.forNodeJS().extractManifest(archive);
    const blobFiles = files.map((f) => ({ path: f.path, content: new Blob([f.content as Buffer]) }));

    return { manifest, files: blobFiles };
  }
}
