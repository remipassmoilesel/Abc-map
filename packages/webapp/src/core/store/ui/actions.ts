/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { HistoryKey } from '../../history/HistoryKey';
import { InformationKey, ServiceWorkerUiState } from './state';

export enum ActionType {
  SetHistoryCapabilities = 'SetHistoryCapabilities',
  CleanHistoryCapabilities = 'CleanHistoryCapabilities',
  SetSideMenuState = 'SetSideMenuState',
  AckInformation = 'AckInformation',
  SetExperimentalFeature = 'SetExperimentalFeature',
  IncrementVisitCounter = 'IncrementVisitCounter',
  SetRemoteModuleUrls = 'SetRemoteModuleUrls',
  RegisterModuleUsage = 'RegisterModuleUsage',
  MarkFavorite = 'MarkFavorite',
  RestoreDefaultFavoriteModules = 'RestoreDefaultFavoriteModules',
  SetServiceWorkerState = 'SetServiceWorkerState',
}

export interface SetHistoryCapabilities {
  type: ActionType.SetHistoryCapabilities;
  key: HistoryKey;
  canUndo: boolean;
  canRedo: boolean;
}

export interface CleanHistoryCapabilities {
  type: ActionType.CleanHistoryCapabilities;
}

export interface SetSideMenuState {
  type: ActionType.SetSideMenuState;
  menuId: string;
  state: boolean;
}

export interface AckInformation {
  type: ActionType.AckInformation;
  name: InformationKey;
}

export interface SetExperimentalFeature {
  type: ActionType.SetExperimentalFeature;
  id: string;
  state: boolean;
}

export interface IncrementVisitCounter {
  type: ActionType.IncrementVisitCounter;
}

export interface SetRemoteModuleUrls {
  type: ActionType.SetRemoteModuleUrls;
  moduleUrls: string[];
}

export interface RegisterModuleUsage {
  type: ActionType.RegisterModuleUsage;
  moduleId: string;
}

export interface MarkFavorite {
  type: ActionType.MarkFavorite;
  moduleId: string;
  favorite: boolean;
}

export interface RestoreDefaultFavoriteModules {
  type: ActionType.RestoreDefaultFavoriteModules;
}

export interface SetServiceWorkerState {
  type: ActionType.SetServiceWorkerState;
  state: Partial<ServiceWorkerUiState>;
}

export type UiAction =
  | SetHistoryCapabilities
  | CleanHistoryCapabilities
  | SetSideMenuState
  | AckInformation
  | IncrementVisitCounter
  | SetExperimentalFeature
  | SetRemoteModuleUrls
  | RegisterModuleUsage
  | MarkFavorite
  | RestoreDefaultFavoriteModules
  | SetServiceWorkerState;

export class UiActions {
  public static setHistoryCapabilities(key: HistoryKey, canUndo: boolean, canRedo: boolean): UiAction {
    return {
      type: ActionType.SetHistoryCapabilities,
      key,
      canUndo,
      canRedo,
    };
  }

  public static cleanHistoryCapabilities(): UiAction {
    return {
      type: ActionType.CleanHistoryCapabilities,
    };
  }

  public static setSideMenuState(menuId: string, state: boolean): UiAction {
    return {
      type: ActionType.SetSideMenuState,
      menuId,
      state,
    };
  }

  public static ackInformation(name: InformationKey): UiAction {
    return {
      type: ActionType.AckInformation,
      name,
    };
  }

  public static setExperimentalFeature(id: string, state: boolean): UiAction {
    return {
      type: ActionType.SetExperimentalFeature,
      id,
      state,
    };
  }

  public static incrementVisitCounter(): UiAction {
    return {
      type: ActionType.IncrementVisitCounter,
    };
  }

  public static setRemoteModuleUrls(moduleUrls: string[]): UiAction {
    return {
      type: ActionType.SetRemoteModuleUrls,
      moduleUrls,
    };
  }

  public static registerModuleUsage(moduleId: string): UiAction {
    return {
      type: ActionType.RegisterModuleUsage,
      moduleId,
    };
  }

  public static markFavorite(moduleId: string, favorite: boolean): UiAction {
    return {
      type: ActionType.MarkFavorite,
      moduleId,
      favorite,
    };
  }

  public static restoreDefaultFavoriteModules(): UiAction {
    return {
      type: ActionType.RestoreDefaultFavoriteModules,
    };
  }

  public static setServiceWorkerState(state: Partial<ServiceWorkerUiState>): UiAction {
    return {
      type: ActionType.SetServiceWorkerState,
      state,
    };
  }
}
