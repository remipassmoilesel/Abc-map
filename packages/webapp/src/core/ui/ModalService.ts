/**
 * Copyright ¬© 2021 R√©mi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Logger } from '@abc-map/shared';
import {
  ConfirmationClosedEvent,
  CreatePasswordModalClosedEvent,
  EditPropertiesClosedEvent,
  LoginClosedEvent,
  ModalEvent,
  ModalEventListener,
  ModalEventType,
  ModalStatus,
  OperationStatus,
  PasswordInputClosedEvent,
  PromptVariablesClosed,
  PwaInstallClosed,
  RegistrationClosedEvent,
  SimplePromptClosed,
  SolicitationClosedEvent,
  TextFeedbackClosed,
} from './typings';
import { DataPropertiesMap } from '../geo/features/FeatureWrapper';
import { resolveInAtLeast } from '../utils/resolveInAtLeast';
import { prefixedTranslation } from '../../i18n/i18n';
import { PromptDefinition } from './PromptDefinition';
import EventEmitter from 'eventemitter3';

const logger = Logger.get('ModalService.ts', 'warn');

const t = prefixedTranslation('ModalService:');

export class ModalService {
  private emitter = new EventEmitter();

  public prompt(title: string, message: string, value: string, validationRegexp: RegExp, validationErrorMessage: string): Promise<SimplePromptClosed> {
    const input: ModalEvent = { type: ModalEventType.ShowSimplePrompt, title, message, value, validationRegexp, validationErrorMessage };
    return this.modalPromise<SimplePromptClosed>(input, ModalEventType.SimplePromptClosed);
  }

  public confirmation(title: string, message: string): Promise<ModalStatus> {
    const input: ModalEvent = { type: ModalEventType.ShowConfirmation, title, message };
    return this.modalPromise<ConfirmationClosedEvent>(input, ModalEventType.ConfirmationClosed).then((res) => res.status);
  }

  /**
   * Show a modal with two password input fields, allowing the user to create a new password
   * @param title
   * @param message
   */
  public createPassword(title: string, message: string): Promise<CreatePasswordModalClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowSetPassword, title, message }, ModalEventType.CreatePasswordClosed);
  }

  public promptPassword(title: string, message: string, witness: string): Promise<PasswordInputClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowPasswordInput, title, message, witness }, ModalEventType.PasswordPromptClosed);
  }

  public editPropertiesModal(properties: DataPropertiesMap): Promise<EditPropertiesClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowEditProperties, properties }, ModalEventType.FeaturePropertiesClosed);
  }

  public solicitation(): Promise<SolicitationClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowSolicitation }, ModalEventType.SolicitationClosed);
  }

  public login(): Promise<LoginClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowLogin }, ModalEventType.LoginClosed);
  }

  public registration(): Promise<RegistrationClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowRegistration }, ModalEventType.RegistrationClosed);
  }

  public passwordLost(): Promise<RegistrationClosedEvent> {
    return this.modalPromise({ type: ModalEventType.ShowPasswordLost }, ModalEventType.PasswordLostClosed);
  }

  public textFeedback(): Promise<TextFeedbackClosed> {
    return this.modalPromise({ type: ModalEventType.ShowTextFeedback }, ModalEventType.TextFeedbackClosed);
  }

  public pwaInstall(): Promise<PwaInstallClosed> {
    return this.modalPromise({ type: ModalEventType.ShowPwaInstall }, ModalEventType.PwaInstallClosed);
  }

  /**
   * Sometimes it is necessary to block the whole UI while an operation is in progress.
   *
   * You can use this method to block UI while returned promise does not resolve.
   *
   * This is generally a bad UX practice.
   *
   * @param operation
   */
  public longOperationModal<Result = void>(operation: () => Promise<OperationStatus | [OperationStatus, Result]>): Promise<OperationStatus | Result> {
    // We show waiting screen
    this.dispatch({ type: ModalEventType.ShowLongOperationModal, processing: true });

    return new Promise<OperationStatus | Result>((resolve, reject) => {
      const start = () => resolveInAtLeast(operation(), 1000).then(hideOnSuccess).catch(hideOnFail);

      const hideOnSuccess = (_result: OperationStatus | [OperationStatus, Result]) => {
        const status = Array.isArray(_result) ? _result[0] : _result;
        const result = Array.isArray(_result) ? _result[1] : _result;

        // Operation was interrupted (per example canceled)
        if (status === OperationStatus.Interrupted) {
          this.dispatch({ type: ModalEventType.LongOperationModalClosed });
          resolve(result ?? OperationStatus.Interrupted);
          return;
        }

        // We show end of waiting screen
        this.dispatch({ type: ModalEventType.ShowLongOperationModal, processing: false });

        setTimeout(() => {
          this.dispatch({ type: ModalEventType.LongOperationModalClosed });

          // Then we resolve with result or status
          resolve(result ?? OperationStatus.Succeed);
        }, 800);
      };

      const hideOnFail = (err: Error) => {
        this.dispatch({ type: ModalEventType.LongOperationModalClosed });
        reject(err);
      };

      // We let frontend refresh before operation
      setTimeout(start, 150);
    });
  }

  public modificationsLostConfirmation(): Promise<ModalStatus> {
    const input: ModalEvent = {
      type: ModalEventType.ShowConfirmation,
      title: `${t('Current_changes')} ‚úçÔ∏è`,
      message: `
          <div class='my-3'>${t('Current_changes_will_be_lost')}</div>
          <div class='my-3'>${t('If_project_saved_ignore_this_message')}</div>
      `,
    };
    return this.modalPromise<ConfirmationClosedEvent>(input, ModalEventType.ConfirmationClosed).then((res) => res.status);
  }

  public dataSizeWarning(): Promise<ModalStatus> {
    const input: ModalEvent = {
      type: ModalEventType.ShowConfirmation,
      title: `${t('Heavy_dataset')} üèãÔ∏è`,
      message: `
          <div class='my-3'>${t('You_are_about_to_import_heavy_dataset')}</div>
          <div class='my-3'>${t('AbcMap_not_yet_ready_for_heavy_dataset')}</div>
          <div class='my-3'>${t('Prefer_small_dataset')}</div>
      `,
    };
    return this.modalPromise<ConfirmationClosedEvent>(input, ModalEventType.ConfirmationClosed).then((res) => res.status);
  }

  public promptVariables(title: string, message: string, definitions: PromptDefinition[]): Promise<PromptVariablesClosed> {
    const input: ModalEvent = { type: ModalEventType.ShowPromptVariables, title, message, definitions };
    return this.modalPromise<PromptVariablesClosed>(input, ModalEventType.PromptVariablesClosed);
  }

  private modalPromise<O extends ModalEvent>(input: ModalEvent, closeEventType: ModalEventType): Promise<O> {
    return new Promise<O>((resolve) => {
      const listener: ModalEventListener = (ev) => {
        if (ev.type === closeEventType) {
          this.removeListener(closeEventType, listener);
          resolve(ev as O);
        } else {
          logger.warn('Unhandled modal event: ', ev);
        }
      };

      this.addListener(closeEventType, listener);
      this.dispatch(input);
    });
  }

  public addListener<T extends ModalEvent = ModalEvent>(type: ModalEventType, listener: ModalEventListener<T>) {
    this.emitter.addListener(type, listener);
  }

  public removeListener(type: ModalEventType, listener: ModalEventListener<any>) {
    this.emitter.removeListener(type, listener);
  }

  public dispatch(event: ModalEvent): void {
    logger.debug('Dispatch event: ', event);
    this.emitter.emit(event.type, event);
  }
}
