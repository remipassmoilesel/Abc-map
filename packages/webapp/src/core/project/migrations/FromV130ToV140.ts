/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcFile, AbcLayer, AbcProjectManifest, LayerType, Logger, PredefinedLayerModel } from '@abc-map/shared';
import { MigrationProject, ProjectMigration, ProjectMigrationSettings } from './typings';
import semver from 'semver';
import { ModalService } from '../../ui/ModalService';
import { AbcProjectManifest130 } from './dependencies/130-project-types';
import { prefixedTranslation } from '../../../i18n/i18n';

const NEXT = '1.4.0';

const logger = Logger.get('FromV130ToV140.ts');

const deprecatedModels = ['StamenToner', 'StamenTonerLite', 'StamenTerrain', 'StamenWatercolor'];

const t = prefixedTranslation('FromV130ToV140:');

/**
 * This migration replaces Stamen predefined layers by OSM layers.
 */
export class FromV130ToV140 implements ProjectMigration<AbcProjectManifest130, AbcProjectManifest> {
  constructor(private modals: ModalService) {}

  public async interestedBy(manifest: AbcProjectManifest130): Promise<boolean> {
    const version = manifest.metadata.version;
    return semver.lt(version, NEXT);
  }

  public async migrate(
    manifest: AbcProjectManifest130,
    files: AbcFile<Blob>[],
    settings?: ProjectMigrationSettings
  ): Promise<MigrationProject<AbcProjectManifest>> {
    // We will warn user once
    let warned = false;
    const layers: AbcLayer[] = manifest.layers.map((layer) => {
      const layerIsDeprecated = LayerType.Predefined === layer.type && deprecatedModels.includes(layer.metadata.model);
      if (layerIsDeprecated) {
        if (!warned && !settings?.silent) {
          this.modals
            .warning(
              t('Changes_have_been_made_to_your_project'),
              `<div class="mb-4">${t('Stamen_layers_are_no_longer_available_as_before')}</div>
               <div class="fw-bold">${t('The_Stamen_layers_in_your_project_have_been_replaced')}</div>`
            )
            .catch((err) => logger.error('Modal error: ', err));
          warned = true;
        }

        return {
          ...layer,
          metadata: {
            ...layer.metadata,
            model: PredefinedLayerModel.OSM,
          },
        };
      }

      return layer;
    });

    const result: MigrationProject<AbcProjectManifest> = {
      manifest: {
        ...manifest,
        metadata: {
          ...manifest.metadata,
          version: NEXT,
        },
        layers,
      },
      files,
    };

    return result;
  }
}
