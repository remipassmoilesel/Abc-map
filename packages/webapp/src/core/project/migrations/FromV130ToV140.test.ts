/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { TestData } from './test-data/TestData';
import { FromV130ToV140 } from './FromV130ToV140';
import { SinonStubbedInstance } from 'sinon';
import { ModalService } from '../../ui/ModalService';
import * as sinon from 'sinon';

describe('FromV130ToV140', () => {
  let modals: SinonStubbedInstance<ModalService>;
  let migration: FromV130ToV140;

  beforeEach(async () => {
    modals = sinon.createStubInstance(ModalService);
    migration = new FromV130ToV140(modals);
  });

  it('interestedBy() should return true if version < 1.4.0', async () => {
    const sampleProject = await TestData.project130();

    expect(await migration.interestedBy(sampleProject.manifest)).toEqual(true);
    expect(await migration.interestedBy(TestData.fakeProject('1.4.0'))).toEqual(false);
  });

  it('should migrate and warn only once', async () => {
    // Prepare
    const sampleProject = await TestData.project130();
    modals.warning.resolves();

    // Act
    const result = await migration.migrate(sampleProject.manifest, sampleProject.files);
    const manifest = result.manifest;

    // Assert
    expect(manifest).toMatchSnapshot();
    expect(modals.warning.callCount).toEqual(1);
    expect(modals.warning.args[0][0]).toEqual('Changes have been made to your project');
  });

  it('should not warn if silent', async () => {
    // Prepare
    const sampleProject = await TestData.project130();
    modals.warning.resolves();

    // Act
    const result = await migration.migrate(sampleProject.manifest, sampleProject.files, { silent: true });
    const manifest = result.manifest;

    // Assert
    expect(manifest).toMatchSnapshot();
    expect(modals.warning.callCount).toEqual(0);
  });
});
