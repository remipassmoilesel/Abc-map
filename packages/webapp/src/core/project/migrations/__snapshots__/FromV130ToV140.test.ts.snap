// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromV130ToV140 should migrate and warn only once 1`] = `
Object {
  "layers": Array [
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "6263689c-dab9-4191-a056-0ae86cd39b95",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "08c4e2f1-dab9-4191-a056-0ae86cd39b95",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "08c4e2f1-dab9-4191-a056-2ae1a6381186",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "features": Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                -456128.347198193,
                6475397.832428156,
              ],
              "type": "Point",
            },
            "id": "P5YHbaiYDE",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -471123.821926691,
                6452071.538406049,
              ],
              "type": "Point",
            },
            "id": "wQKWYEr5-B",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -449463.691763305,
                6430411.408242662,
              ],
              "type": "Point",
            },
            "id": "6wPiMcLpa5",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      },
      "metadata": Object {
        "active": true,
        "id": "08c4e2f1-715a-4d90-8c33-2ae1a6381186",
        "name": "Geometries",
        "opacity": 1,
        "type": "Vector",
        "visible": true,
      },
      "type": "Vector",
    },
  ],
  "layouts": Object {
    "abcMapAttributionsEnabled": true,
    "list": Array [],
  },
  "metadata": Object {
    "id": "7f6381eb-3f87-4893-870a-e72e85679320",
    "name": "Projet du 10/3/2023",
    "public": false,
    "version": "1.4.0",
  },
  "sharedViews": Object {
    "fullscreen": false,
    "list": Array [],
    "mapDimensions": Object {
      "height": 601,
      "width": 1151,
    },
  },
  "view": Object {
    "center": Array [
      -470972.352484989,
      6429426.856871599,
    ],
    "projection": Object {
      "name": "EPSG:3857",
    },
    "resolution": 151.469441702,
    "rotation": 0,
  },
}
`;

exports[`FromV130ToV140 should not warn if silent 1`] = `
Object {
  "layers": Array [
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "6263689c-dab9-4191-a056-0ae86cd39b95",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "08c4e2f1-dab9-4191-a056-0ae86cd39b95",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "metadata": Object {
        "active": false,
        "attributions": Array [
          "Map tiles by <a href=\\"https://stamen.com/\\" target=\\"_blank\\">Stamen Design</a>, under <a href=\\"https://creativecommons.org/licenses/by/3.0/\\" target=\\"_blank\\">CC BY 3.0</a>.",
          "&#169; <a href=\\"https://www.openstreetmap.org/copyright\\" target=\\"_blank\\">OpenStreetMap</a> contributors.",
        ],
        "id": "08c4e2f1-dab9-4191-a056-2ae1a6381186",
        "model": "OSM",
        "name": "Stamen Watercolor",
        "opacity": 1,
        "type": "Predefined",
        "visible": true,
      },
      "type": "Predefined",
    },
    Object {
      "features": Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                -456128.347198193,
                6475397.832428156,
              ],
              "type": "Point",
            },
            "id": "P5YHbaiYDE",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -471123.821926691,
                6452071.538406049,
              ],
              "type": "Point",
            },
            "id": "wQKWYEr5-B",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -449463.691763305,
                6430411.408242662,
              ],
              "type": "Point",
            },
            "id": "6wPiMcLpa5",
            "properties": Object {
              "abc:feature:selected": true,
              "abc:style:point:color": "rgba(18,90,147,0.9)",
              "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
              "abc:style:point:size": 30,
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      },
      "metadata": Object {
        "active": true,
        "id": "08c4e2f1-715a-4d90-8c33-2ae1a6381186",
        "name": "Geometries",
        "opacity": 1,
        "type": "Vector",
        "visible": true,
      },
      "type": "Vector",
    },
  ],
  "layouts": Object {
    "abcMapAttributionsEnabled": true,
    "list": Array [],
  },
  "metadata": Object {
    "id": "7f6381eb-3f87-4893-870a-e72e85679320",
    "name": "Projet du 10/3/2023",
    "public": false,
    "version": "1.4.0",
  },
  "sharedViews": Object {
    "fullscreen": false,
    "list": Array [],
    "mapDimensions": Object {
      "height": 601,
      "width": 1151,
    },
  },
  "view": Object {
    "center": Array [
      -470972.352484989,
      6429426.856871599,
    ],
    "projection": Object {
      "name": "EPSG:3857",
    },
    "resolution": 151.469441702,
    "rotation": 0,
  },
}
`;
